### ISSUE TITLE
How can I use this without TLS?

### ISSUE THREAD

**Issue by doccccccccc** [omjadas/hudsucker#46]
I want to proxy http only requests. Is this possible?

**Comment by omjadas** [comment]
When you say use without TLS, do you mean block any attempts to establish a TLS connection, or do you mean tunnel TLS connections without intercepting?

**Comment by doccccccccc** [comment]
Running it without a root certificate

**Comment by omjadas** [comment]
I have published [v0.19.1](https://github.com/omjadas/hudsucker/releases/tag/v0.19.1), which allows the `HttpHandler` to control whether a `CONNECT` request should be intercepted. If you do something like the following (where you always return `false` from `should_intercept`) then the CA will never be used and thus doesn't need to do anything. Of course, this means that you won't be able to intercept HTTPS requests.

```rust
use http::uri::Authority;
use hudsucker::{
    async_trait::async_trait,
    certificate_authority::CertificateAuthority,
    hyper::{Body, Client, Request},
    rustls::ServerConfig,
    *,
};
use std::{net::SocketAddr, sync::Arc};

async fn shutdown_signal() {
    tokio::signal::ctrl_c()
        .await
        .expect("Failed to install CTRL+C signal handler");
}

struct NoCa;

#[async_trait]
impl CertificateAuthority for NoCa {
    async fn gen_server_config(&self, _authority: &Authority) -> Arc<ServerConfig> {
        unreachable!();
    }
}

#[derive(Clone)]
struct MyHandler;

#[async_trait]
impl HttpHandler for MyHandler {
    async fn should_intercept(&mut self, _ctx: &HttpContext, _req: &Request<Body>) -> bool {
        false
    }
}

#[tokio::main]
async fn main() {
    let proxy = Proxy::builder()
        .with_addr(SocketAddr::from(([127, 0, 0, 1], 3000)))
        .with_client(Client::new())
        .with_ca(NoCa)
        .with_http_handler(MyHandler)
        .build();

    proxy.start(shutdown_signal()).await.unwrap()
}
```

**Comment by bbaldino** [comment]
@omjadas would it be possible to enable a mode where there's no certificate exchange between hudsucker and the client, but still allow for intercepting https?  I've got a scenario where the link between the client and the proxy has no need for https (and a big goal is to alleviate the client from having to deal with https at all), but the proxy itself would handle https connections and certificate verification on that "side".

Basically, I want client <-> hudsucker to be plain http, even in scenarios where the client is accessing an https url via hudsucker, but still need to be able to intercept those (https) requests in hudsucker (so return false in `should_intercept` wouldn't work).

**Comment by omjadas** [comment]
@bbaldino that is not possible.

**Comment by omjadas** [comment]
> Oh alright. Why is that? I'm curious to understand. Is it a limitation of the proxy "spec" in some way? Or the hudsucker code?

It is due to the way that HTTP proxies work. I would recommend looking into [RFC 2817](https://datatracker.ietf.org/doc/html/rfc2817) for details.
